File: apps/macos/Sources/Overnight/Markdown.swift
import SwiftUI

/// Markdown, rendered as blocks rather than as one run of text.
///
/// `AttributedString(markdown:)` handles inline syntax — bold, italic, code
/// spans, links — and nothing else. Headings arrive as plain text, list items
/// lose their bullets, and every block is concatenated into a single
/// paragraph, which is why a reply full of structure rendered as one long
/// line with some words emphasised.
///
/// So blocks are split here and each is drawn as itself; inline parsing is
/// still `AttributedString`'s job, which it does well. This is not a complete
/// CommonMark implementation and is not trying to be — it covers what a chat
/// reply actually contains.
enum Markdown {
    /// One renderable piece of a reply.
    enum Block: Equatable {
        case paragraph(String)
        case heading(level: Int, text: String)
        case bullet(text: String, depth: Int)
        case numbered(number: String, text: String, depth: Int)
        case code(text: String, language: String)
        case quote(String)
        case rule
    }

    /// Split markdown into blocks.
    ///
    /// Line-based, because the structures that matter here are line-based. A
    /// fenced code block swallows everything until its closing fence, so its
    /// contents are never mistaken for markup.
    static func blocks(_ text: String) -> [Block] {
        var blocks: [Block] = []
        var paragraph: [String] = []

        func flushParagraph() {
            // Joined with a NEWLINE, not a space.
            //
            // CommonMark folds a single line break into a space and needs two
            // to make one. That rule exists for hand-written source files, and
            // Claude does not write to it — it breaks lines where it means
            // them to break. Honouring that is the difference between a
            // readable reply and one run-on paragraph.
            let joined = paragraph.joined(separator: "\n").trimmingCharacters(in: .whitespaces)
            if !joined.isEmpty { blocks.append(.paragraph(joined)) }
            paragraph = []
        }

        var lines = text.components(separatedBy: .newlines)[...]
        while let line = lines.first {
            lines = lines.dropFirst()
            let trimmed = line.trimmingCharacters(in: .whitespaces)

            // A fence takes precedence over everything: its contents are not
            // markup and must not be read as any.
            if trimmed.hasPrefix("```") {
                flushParagraph()
                let language = String(trimmed.dropFirst(3)).trimmingCharacters(in: .whitespaces)
                var body: [String] = []
                while let next = lines.first {
                    lines = lines.dropFirst()
                    if next.trimmingCharacters(in: .whitespaces).hasPrefix("```") { break }
                    body.append(next)
                }
                blocks.append(.code(text: body.joined(separator: "\n"), language: language))
                continue
            }

            if trimmed.isEmpty {
                flushParagraph()
                continue
            }

            if trimmed == "---" || trimmed == "***" || trimmed == "___" {
                flushParagraph()
                blocks.append(.rule)
                continue
            }

            if let hashes = trimmed.firstIndex(where: { $0 != "#" }),
                trimmed.starts(with: "#"),
                trimmed[hashes] == " "
            {
                flushParagraph()
                let level = trimmed.distance(from: trimmed.startIndex, to: hashes)
                let body = String(trimmed[hashes...]).trimmingCharacters(in: .whitespaces)
                blocks.append(.heading(level: min(level, 3), text: body))
                continue
            }

            if trimmed.hasPrefix("> ") {
                flushParagraph()
                blocks.append(.quote(String(trimmed.dropFirst(2))))
                continue
            }

            // Indentation is what nests a list, so it is measured before the
            // marker is stripped.
            let indent = line.prefix(while: { $0 == " " || $0 == "\t" }).count
            let depth = min(indent / 2, 3)

            if let marker = ["- ", "* ", "+ "].first(where: { trimmed.hasPrefix($0) }) {
                flushParagraph()
                blocks.append(.bullet(text: String(trimmed.dropFirst(marker.count)), depth: depth))
                continue
            }

            if let dot = trimmed.firstIndex(of: "."),
                trimmed[trimmed.startIndex..<dot].allSatisfy(\.isNumber),
                trimmed.index(after: dot) < trimmed.endIndex,
                trimmed[trimmed.index(after: dot)] == " "
            {
                flushParagraph()
                let number = String(trimmed[trimmed.startIndex..<dot])
                let body = String(trimmed[trimmed.index(dot, offsetBy: 2)...])
                blocks.append(.numbered(number: number, text: body, depth: depth))
                continue
            }

            paragraph.append(trimmed)
        }
        flushParagraph()
        return blocks
    }

    /// Inline syntax only — bold, italic, code spans, links.
    ///
    /// Text that cannot be parsed is returned as itself rather than dropped: a
    /// stray bracket should cost a reader the emphasis, not the sentence.
    static func inline(_ text: String) -> AttributedString {
        let options = AttributedString.MarkdownParsingOptions(
            allowsExtendedAttributes: true,
            interpretedSyntax: .inlineOnlyPreservingWhitespace,
            failurePolicy: .returnPartiallyParsedIfPossible
        )
        return (try? AttributedString(markdown: text, options: options)) ?? AttributedString(text)
    }
}

/// A rendered markdown reply.
struct MarkdownText: View {
    let text: String
    /// Reasoning is set smaller and dimmer than a reply, but is otherwise the
    /// same markdown — agents write lists and code in their thinking too.
    var secondary: Bool = false

    var body: some View {
        VStack(alignment: .leading, spacing: 7) {
            ForEach(Array(Markdown.blocks(text).enumerated()), id: \.offset) { _, block in
                view(for: block)
            }
        }
        .font(secondary ? .caption : .body)
        .foregroundStyle(secondary ? AnyShapeStyle(.secondary) : AnyShapeStyle(.primary))
        .textSelection(.enabled)
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    @ViewBuilder
    private func view(for block: Markdown.Block) -> some View {
        switch block {
        case let .paragraph(text):
            Text(Markdown.inline(text))
                .fixedSize(horizontal: false, vertical: true)

        case let .heading(level, text):
            Text(Markdown.inline(text))
                .font(headingFont(level))
                .fixedSize(horizontal: false, vertical: true)
                .padding(.top, 2)

        case let .bullet(text, depth):
            marker("•", text: text, depth: depth)

        case let .numbered(number, text, depth):
            marker("\(number).", text: text, depth: depth)

        case let .code(text, _):
            // The same box a tool's output gets: one way of showing
            // monospaced text, not two.
            DetailBox(text: text)

        case let .quote(text):
            HStack(spacing: 8) {
                Rectangle().fill(.quaternary).frame(width: 2)
                Text(Markdown.inline(text))
                    .fixedSize(horizontal: false, vertical: true)
            }

        case .rule:
            Divider()
        }
    }

    /// A marker and its text, aligned so a wrapped line does not slide back
    /// under the bullet.
    private func marker(_ symbol: String, text: String, depth: Int) -> some View {
        HStack(alignment: .firstTextBaseline, spacing: 6) {
            Text(symbol)
                .foregroundStyle(.secondary)
                .frame(minWidth: 14, alignment: .trailing)
            Text(Markdown.inline(text))
                .fixedSize(horizontal: false, vertical: true)
            Spacer(minLength: 0)
        }
        .padding(.leading, CGFloat(depth) * 14)
    }

    private func headingFont(_ level: Int) -> Font {
        switch level {
        case 1: secondary ? .callout.bold() : .title3.bold()
        case 2: secondary ? .caption.bold() : .headline
        default: secondary ? .caption.bold() : .subheadline.bold()
        }
    }
}

/// The app's motion, in one place.
///
/// Springs rather than timed curves, and short ones. A disclosure that takes
/// a third of a second to open reads as the app thinking about it; the whole
/// point of a fold is that it costs nothing to look.
enum Motion {
    /// Opening, closing, and anything else that should feel instant.
    static let snap = Animation.spring(response: 0.22, dampingFraction: 0.82)
    /// A little overshoot, for something arriving rather than resizing.
    static let arrive = Animation.spring(response: 0.26, dampingFraction: 0.7)
}

File: apps/android/app/src/main/java/com/farcooler/model/Markdown.kt
package com.farcooler.model

/**
 * Markdown, split into the pieces a reply is drawn from.
 *
 * Ported from `apps/shared/AgentKit/Sources/AgentKit/MarkdownView.swift`. Only
 * the parser is here; the drawing is in the UI layer, because layout is each
 * platform's own job and a block list is not.
 *
 * Line-based, because the structures that matter here are line-based. A fenced
 * code block swallows everything until its closing fence, so its contents are
 * never mistaken for markup.
 */
object Markdown {
    /** One renderable piece of a reply. */
    sealed interface Block {
        data class Paragraph(val text: String) : Block
        data class Heading(val level: Int, val text: String) : Block
        data class Bullet(val text: String, val depth: Int) : Block
        data class Numbered(val number: String, val text: String, val depth: Int) : Block
        data class Code(val text: String, val language: String) : Block
        data class Quote(val text: String) : Block
        data object Rule : Block

        /**
         * A pipe table. The header row is kept apart from the body because it
         * is drawn differently, not because the parser needs it separated.
         */
        data class Table(val header: List<String>, val rows: List<List<String>>) : Block
    }

    /**
     * Split one `| a | b |` line into its cells.
     *
     * The outer pipes are optional in GitHub's dialect and agents write them
     * either way, so an empty cell from a leading or trailing pipe is dropped
     * rather than rendered as a blank column.
     */
    fun cells(line: String): List<String> {
        var trimmed = line.trim()
        if (trimmed.startsWith("|")) trimmed = trimmed.drop(1)
        if (trimmed.endsWith("|")) trimmed = trimmed.dropLast(1)
        return trimmed.split("|").map { it.trim() }
    }

    /**
     * Whether a line is a table's `|---|:--:|` separator.
     *
     * This is what distinguishes a table from a paragraph that happens to
     * contain a pipe — a shell command, most often, which must not be eaten as
     * markup.
     */
    fun isTableRule(line: String): Boolean {
        if (!line.contains("|")) return false
        val parts = cells(line)
        if (parts.isEmpty()) return false
        return parts.all { cell ->
            cell.isNotEmpty() && cell.all { it == '-' || it == ':' || it == ' ' } && cell.contains('-')
        }
    }

    fun blocks(text: String): List<Block> {
        val blocks = mutableListOf<Block>()
        val paragraph = mutableListOf<String>()

        fun flushParagraph() {
            // Joined with a NEWLINE, not a space.
            //
            // CommonMark folds a single line break into a space and needs two
            // to make one. That rule exists for hand-written source files, and
            // an agent does not write to it — it breaks lines where it means
            // them to break. Honouring that is the difference between a
            // readable reply and one run-on paragraph.
            val joined = paragraph.joinToString("\n").trim()
            if (joined.isNotEmpty()) blocks.add(Block.Paragraph(joined))
            paragraph.clear()
        }

        val lines = text.split("\n")
        var index = 0
        while (index < lines.size) {
            val line = lines[index]
            index += 1
            val trimmed = line.trim()

            // A fence takes precedence over everything: its contents are not
            // markup and must not be read as any.
            if (trimmed.startsWith("```")) {
                flushParagraph()
                val language = trimmed.drop(3).trim()
                val body = mutableListOf<String>()
                while (index < lines.size) {
                    val next = lines[index]
                    index += 1
                    if (next.trim().startsWith("```")) break
                    body.add(next)
                }
                blocks.add(Block.Code(body.joinToString("\n"), language))
                continue
            }

            if (trimmed.isEmpty()) {
                flushParagraph()
                continue
            }

            if (trimmed == "---" || trimmed == "***" || trimmed == "___") {
                flushParagraph()
                blocks.add(Block.Rule)
                continue
            }

            if (trimmed.startsWith("#")) {
                val level = trimmed.takeWhile { it == '#' }.length
                if (level < trimmed.length && trimmed[level] == ' ') {
                    flushParagraph()
                    blocks.add(Block.Heading(minOf(level, 3), trimmed.drop(level).trim()))
                    continue
                }
            }

            // A table is recognised by its SECOND line, not its first: the
            // header alone is indistinguishable from a sentence containing a
            // pipe. So the separator is peeked at before either is consumed.
            if (trimmed.contains("|") && index < lines.size && isTableRule(lines[index])) {
                flushParagraph()
                index += 1
                val header = cells(trimmed)
                val body = mutableListOf<List<String>>()
                while (index < lines.size) {
                    val row = lines[index].trim()
                    if (row.isEmpty() || !row.contains("|")) break
                    index += 1
                    body.add(cells(row))
                }
                blocks.add(Block.Table(header, body))
                continue
            }

            if (trimmed.startsWith("> ")) {
                flushParagraph()
                blocks.add(Block.Quote(trimmed.drop(2)))
                continue
            }

            // Indentation is what nests a list, so it is measured before the
            // marker is stripped.
            val indent = line.takeWhile { it == ' ' || it == '\t' }.length
            val depth = minOf(indent / 2, 3)

            val marker = listOf("- ", "* ", "+ ").firstOrNull { trimmed.startsWith(it) }
            if (marker != null) {
                flushParagraph()
                blocks.add(Block.Bullet(trimmed.drop(marker.length), depth))
                continue
            }

            val dot = trimmed.indexOf('.')
            if (dot > 0 &&
                trimmed.take(dot).all { it.isDigit() } &&
                dot + 1 < trimmed.length &&
                trimmed[dot + 1] == ' '
            ) {
                flushParagraph()
                blocks.add(Block.Numbered(trimmed.take(dot), trimmed.drop(dot + 2), depth))
                continue
            }

            paragraph.add(trimmed)
        }
        flushParagraph()
        return blocks
    }

    /** One run of inline text, with whatever emphasis applies to it. */
    data class Span(
        val text: String,
        val bold: Boolean = false,
        val italic: Boolean = false,
        val code: Boolean = false,
        val link: String? = null,
    )

    /**
     * Inline syntax only — bold, italic, code spans, links.
     *
     * Hand-rolled, unlike Swift's, which hands the string to Foundation's own
     * markdown parser. Android has no equivalent, and pulling in a markdown
     * library to resolve four constructs would be a large dependency for a
     * small job.
     *
     * Text that cannot be parsed is returned as itself rather than dropped: a
     * stray bracket should cost a reader the emphasis, not the sentence. Code
     * spans are resolved first and their contents are never re-scanned, so an
     * asterisk inside `**` in a shell command stays an asterisk.
     */
    fun inline(text: String): List<Span> {
        val spans = mutableListOf<Span>()
        val literal = StringBuilder()

        fun flush() {
            if (literal.isNotEmpty()) {
                spans.add(Span(literal.toString()))
                literal.clear()
            }
        }

        var i = 0
        while (i < text.length) {
            val rest = text.substring(i)

            // Backticks first, and their contents are taken verbatim.
            if (rest.startsWith("`")) {
                val end = rest.indexOf('`', startIndex = 1)
                if (end > 0) {
                    flush()
                    spans.add(Span(rest.substring(1, end), code = true))
                    i += end + 1
                    continue
                }
            }

            if (rest.startsWith("**") || rest.startsWith("__")) {
                val fence = rest.take(2)
                val end = rest.indexOf(fence, startIndex = 2)
                if (end > 0) {
                    flush()
                    spans.add(Span(rest.substring(2, end), bold = true))
                    i += end + 2
                    continue
                }
            }

            if ((rest.startsWith("*") || rest.startsWith("_")) && rest.length > 1) {
                val fence = rest.take(1)
                val end = rest.indexOf(fence, startIndex = 1)
                if (end > 1) {
                    flush()
                    spans.add(Span(rest.substring(1, end), italic = true))
                    i += end + 1
                    continue
                }
            }

            if (rest.startsWith("[")) {
                val close = rest.indexOf(']')
                if (close > 0 && close + 1 < rest.length && rest[close + 1] == '(') {
                    val end = rest.indexOf(')', startIndex = close + 2)
                    if (end > 0) {
                        flush()
                        spans.add(
                            Span(
                                rest.substring(1, close),
                                link = rest.substring(close + 2, end),
                            )
                        )
                        i += end + 1
                        continue
                    }
                }
            }

            literal.append(text[i])
            i += 1
        }
        flush()
        return spans
    }
}

/**
 * How far through a plan entry is, as the agent reports it.
 *
 * The wire carries free-form strings from several adapters, so the mapping
 * lives in one place rather than in each surface that draws a checkmark.
 */
enum class PlanStatus {
    PENDING,
    ACTIVE,
    DONE;

    val isDone: Boolean get() = this == DONE

    companion object {
        fun parse(raw: String): PlanStatus = when (raw.lowercase()) {
            "completed", "done" -> DONE
            "in_progress", "inprogress", "active" -> ACTIVE
            else -> PENDING
        }
    }
}

val List<PlanEntry>.doneCount: Int
    get() = count { PlanStatus.parse(it.status) == PlanStatus.DONE }

val List<PlanEntry>.active: PlanEntry?
    get() = firstOrNull { PlanStatus.parse(it.status) == PlanStatus.ACTIVE }
